src/Microsoft.AspNetCore.OData/ODataQueryParameterBindingAttribute.cs

```diff
@@ -2,15 +2,144 @@
-    /// A <see cref="Attribute"/> to bind parameters of type <see cref="ODataQueryOptions"/> to the OData query from the incoming request.
+    /// A <see cref="ModelBinderAttribute"/> to bind parameters of type <see cref="ODataQueryOptions"/> to the OData query from the incoming request.
     /// </summary>
     [AttributeUsage(AttributeTargets.Class | AttributeTargets.Parameter, Inherited = true, AllowMultiple = false)]
-    public sealed partial class ODataQueryParameterBindingAttribute : Attribute
+    public sealed partial class ODataQueryParameterBindingAttribute : ModelBinderAttribute
     {
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="ODataQueryParameterBindingAttribute"/> class.
+        /// </summary>
+        public ODataQueryParameterBindingAttribute()
+            :base(typeof(ODataQueryParameterBinding))
+        {
+        }
+
+        internal class ODataQueryParameterBinding : IModelBinder
+        {
+            private static MethodInfo _createODataQueryOptions = typeof(ODataQueryParameterBinding).GetMethod("CreateODataQueryOptions");
+            private const string CreateODataQueryOptionsCtorKey = "MS_CreateODataQueryOptionsOfT";
+
+            [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "We don't want to fail in model binding.")]
+            public Task BindModelAsync(ModelBindingContext bindingContext)
+            {
+                if (bindingContext == null)
+                {
+                    throw Error.ArgumentNull("bindingContext");
+                }
+
+                HttpRequest request = bindingContext.HttpContext.Request;
+
+                if (request == null)
+                {
+                    throw Error.Argument("actionContext", SRResources.ActionContextMustHaveRequest);
+                }
+
+                ActionDescriptor actionDescriptor = bindingContext.ActionContext.ActionDescriptor;
+
+                if (actionDescriptor == null)
+                {
+                    throw Error.Argument("actionContext", SRResources.ActionContextMustHaveDescriptor);
+                }
+
+                // Get the entity type from the parameter type if it is ODataQueryOptions<T>.
+                // Fall back to the return type if not. Also, note that the entity type from the return type and ODataQueryOptions<T>
+                // can be different (example implementing $select or $expand).
+                Type entityClrType = null;
+                ParameterDescriptor paramDescriptor = bindingContext.ActionContext.ActionDescriptor.Parameters.FirstOrDefault();
+                if (paramDescriptor != null)
+                {
+                    entityClrType = GetEntityClrTypeFromParameterType(paramDescriptor.ParameterType);
+                }
+
+                if (entityClrType == null)
+                {
+                    entityClrType = GetEntityClrTypeFromActionReturnType(actionDescriptor);
+
```

Comment:
Empty line. Not sure if stylecop catches this.

Reply:
It should, let me look into that...